src/ansys/fluent/core/session_meshing.py

```diff
@@ -66,3 +72,120 @@ def PMFileManagement(self):
         return super(Meshing, self).preferences if not self.switched else None
+
+    def read_case(
+        self,
+        file_name: str,
+    ):
+        """Reads a case file.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        """
+        if not self._server_file_manager:
+            self._server_file_manager = _ServerFileManager(self._fluent_connection)
+        return self._server_file_manager.read_case(
+            file_name,
+        )
+
+    def write_case(
+        self,
+        file_name: str,
+    ):
+        """Writes a case file.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        """
+        if not self._server_file_manager:
+            self._server_file_manager = _ServerFileManager(self._fluent_connection)
+        return self._server_file_manager.write_case(file_name)
+
+
+class _ServerFileManager(Meshing):
+    """Supports file upload and download for every existing file read,
+    write methods respectively in the cloud particularly in Ansys lab.
+    Here we are supporting upload and download methods in existing session
+    methods. These would be no-ops if PyPIM is not configured or not authorized
+    with the appropriate service. This will be used for internal purpose only.
+
+    Methods
+    -------
+    read_case(
+        file_name
+        )
+        Read a case file.
+
+    write_case(
+        file_name
+        )
+        Write a case file.
+    """
+
+    def __init__(self, fluent_connection):
+        super(_ServerFileManager, self).__init__(fluent_connection)
+        self.pim_instance = fluent_connection._remote_instance
+        self.file_service = None
+        try:
+            upload_server = self.pim_instance.services["http-simple-upload-server"]
+        except (AttributeError, KeyError):
+            pass
+        else:
+            self.file_service = Client(
+                token="token", url=upload_server.uri, headers=upload_server.headers
+            )
+
+    def _wait_for_file(self, file_name, file_service):
+        start_time = time.time()
+        max_wait_time = 100
+        while (time.time() - start_time) < max_wait_time:
+            if file_service.file_exist(os.path.basename(file_name)):
+                break
+            max_wait_time -= 1
+            time.sleep(3)
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+
+    def read_case(
+        self,
+        file_name: str,
+    ):
+        """Reads  a case file.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        """
+        if pypim.is_configured():
+            if os.path.isfile(file_name):
+                self.upload(os.path.basename(file_name))
+            else:
+                raise FileNotFoundError(f"{file_name} does not exist.")
+            self._wait_for_file(file_name, self.file_service)
```

Comment:
A slightly different strategy that we discussed is checking if the file exists remotely.

I think it is reasonable and correct to first check local existence, as you are doing. But if it doesn't exist locally, then we could check remote existence directly. This could cover a valid use case.

Can `upload()` fail if the file already exists remotely, and if so, does it raise an exception. If so, that is fine, we should try-except that code, and simply proceed with the `read_case()`.

It would be useful to have a notebook avaialbel to run in the lab and test all these scenarios.

Reply:
Yes, we will implement this check. Thank you.